### ISSUE TITLE
How to run merge multisamples with popins?

### ISSUE THREAD

**Issue by Jordi-V** [bkehr/popins#7]
Dear bkehr,

I want to use popins in my samples. I've 808 samples at 30X, I ran all my samples with popins assemble, and now I want to run popins merge, but I dont know how to do it or how it works?? because I ran the first step like this:
popins assemble -s x -t 48 -mp path/to/bam/x.bam -p output/directory/ x.bam
popins assemble -s x -t 48 -mp path/to/bam/y.bam -p output/directory/ y.bam
popins assemble -s x -t 48 -mp path/to/bam/z.bam -p output/directory/ z.bam
 
All my outputs goes to the output directory and now how I can say merge all my samples??
i write for each sample this...
popins merge -p  output/directory/  -c  output/directory/supercontig_x.fa

but i dont know how to do it for all my samples....
The same happens for step:

popins place commands...

Even more I want to know if is necessary merge all my samples with a coverage of 30X...

an example how run multiple samples will be appreciate.... 

thanks for your help

Jordi

**Comment by bkehr** [comment]
Dear Jordi,

this might be a typo in your message above, but the assemble commands you have specified there will overwrite the sample directory ``output/directory/x/`` again and again. You need to specify unique names for each sample with the ``-s`` option, e.g. ``x``, ``y`` and ``z``.

Once you get unique sample directories, you can check the input for the merge command using ``ls output/directory/*/contigs.fa``. These contig files of all the samples are found and merged by the popdel merge command when specifying ``-p output/directory/`` as you did above.

The idea of PopIns is to generate one set of insertions for a large number of samples. This is why we introduced the merging step. Once the supercontigs are merged and placed, you obtain a single set of insertions for all input samples that you can then genotype in each of the input samples. Merging 808 samples of 30X coverage is possible even though it may require quite a bit of memory. If you do not want to compare insertion calls between samples, then you can of course skip the merging and run the remaining steps per sample.

Hope this helps,

Birte

**Comment by Jordi-V** [comment]
Thanks a lot for your help Birte!

I've a new question about popins Birte...

The insertion.vcf obtained before to genotype this vcf for each sample, I found that the insertion.vcf repeat the same chromosome two times... It means, for example:
1       10004   1:10004:FP      C       ]CWGS8.NODE_60601_length_94_cov_2.000000r:57]C        .       .       AR=15;AS=0.394737;GS=2
1       72091   1:72091:FP      A       ]CWGS3.NODE_69810_length_104_cov_2.086539r:120]A      .       .       NOANCHOR
1       86485   1:86485:FP      T       T[CWGS4.NODE_56398_length_125_cov_2.720000r:107[      .       .       AR=2;AS=1;GS=1
1       99293   1:99293:FP      T       T[CWGS5.NODE_70452_length_139_cov_3.021583f:98[       .       .       AR=59;AS=0.359756;GS=2
1       99228   1:99228:FP      G       ]CWGS7.NODE_75505_length_139_cov_2.992806f:85]G       .       .       NOANCHOR
1       99138   1:99138:FP      C       ]CWGS0.NODE_7742_length_132_cov_3.045455r:67]C        .       .       NOANCHOR
1       243397  1:243397:FP     T       ]CWGS1.NODE_68742_length_140_cov_2.871428r:82]T       .       .       AR=2;AS=0.5;GS=2
1       136911  1:136911:FP     C       C[CWGS9.NODE_53956_length_134_cov_3.134328r:85[       .       .       AR=34;AS=0.435897;GS=10
.
.
hs37d5 ...
hs37d5...
and start again to chr1
1...
1...
1...
to hs37d5

after see all chromosomes (the chr1 to hs37d5), the chromosomes starts again the chromosome 1, where the amount of variants is bigger than the first time when I see the chr1. So I dont understand why happen this... all variants are correct??? first appear all chromosomes with less variants than the second time, the first time when I see the chromosomes are the good ones? 

Thanks for your help

**Comment by bkehr** [comment]
Dear Jordi,

you should consider all variants in the output. The reason, why you see this ordering, is that PopIns uses two approaches for placing contigs: reference alignment and split read alignment. PopIns first attempts to place all contigs using reference alignment. This creates the larger fraction of lines in the VCF file. Secondly, it attempts to place those contigs that it couldn't place in the first round using split read alignment. This creates the second batch of lines in the VCF file.

The genotyping is not influenced by the order of variants, which is why we did not bother to sort the VCF file. You may want to sort it by coordinate for your downstream analyses.

Birte

**Comment by Jordi-V** [comment]
Okay Brite,

so If I understand correct all variants mainly those are not NONANCHOR are good for the further analysis. And If I want I can sorted in order to obtain all variants by chromosome.

One thing which surprise me is the amount of de novo insertions found by popins, for example for one sample manta is able to find 2337 breakpoints and Popins for same sample around 18000 bkp. Is surprisingly the amount of bkp founded by popins. I checked in the bam those bkps and seems that there are bkps...

Thanks a lot for your help Brite

**Comment by bkehr** [comment]
Yes, correct, you can sort the variants by chromosome and position if you need them to be ordered.

PopIns is very generous in what it reports as a breakpoint. Usually, the output contains false positives, which is why we used stringend filtering on the Icelandic call set (see [Kehr et al. 2017](https://www.nature.com/articles/ng.3801)). I recommend to you to apply similar filters.

Good luck with your project!

Birte

**Comment by Jordi-V** [comment]
Dear Brite,

Thanks for your help. I checked the steps used in the paper which you recommended me to follow. In the last release of popins, some flags are not available like (bamCov), On the other hand I follow the steps from github, and only there are two filters which I dont change and the steps from the paper are different. One is the minEntropy, where by default is 0.75 and the paper is 0.88. This difference is high?? On the other hand the --filter in assemble step, I dont include this flag because in the documentation indicates this "Recommended for non-human reference sequences", and I've a human samples... even more I've pared-end bams so I include the -mp flag...

The steps which I include are the following:
popins assemble -s sample -t 24 -mp sample.bam -p all_GCAT sample.bam
popins merge -p all_GCAT -c supercontigs.fa
popins contigmap -r genome.fa -t 48 -p all_GCAT -c supercontigs.fa sample
popins place-refalign -r genome.fa  --readLength 150  -p all_GCAT  -c supercontigs.fa
popins place-splitalign -r genome.fa -p all_GCAT --readLength 150 -c supercontigs.fa sample
popins place-finish -r genome.fa -p all_GCAT  -i insertions.vcf
popins genotype -r genome.fa -p all_GCAT -i insertions.vcf -c supercontigs.fa sample

for genotype I decided put the the option RANDOM in the flag -m, because I dont know the importance to put the DUP... Could you tell me If I can imporve the variant detection by popins?? In order to decrese the flase positive rate??
As I've mentio  before I've a set of 785 samples at 30X, and I analize the recurrences between samples in order to calculate the Allele frequency. In some variants all 785 samples find exactly the same genotype... This is a false positive?? Why in some positions all samples contain the same breakpoint? could be related to HLA regions? These are some positions as example:
Chr  position  genotype MAF frequency
12 81765101 1/1 785 MAF 1.00000
1 144010525 0/1 785 MAF 0.50000
15 55284735 1/1 784 0/1 1 MAF 0.99936
17 8069471 1/1 785 MAF 1.00000

These positions could be false positives?? I check in the bam files these positions and I found breakpoints but I dont know If this positions could be discarded....

THanks again for your help.

JOrdi

**Comment by Jordi-V** [comment]
Hi Brite,

sorry for insistence... But do you think that the pipeline used to run popins is correct?? my previous post show the steps...

Thanks for your time and help

**Comment by bkehr** [comment]
Hi Jordi,

sorry for the late reply, I was travelling when you posted your pipeline commands.

1. We decided to use a more strict entropy filter on the Icelandic data than the default value. Both thresholds will work. The lower you set it the more repetitive sequences (and false positives) you will get.

2. The --filter option in the assemble command is only applicable in combination with the --reference option if you would like to filter out contamination (known human and non-human sequences). If a reference fasta file is specified, the reads are first remapped to this reference using BWA-MEM and only reads that remain without high-quality alignment after remapping are quality-filtered and assembled. For filtering contamination from the Icelandic data, we concatenated alternative haplotype sequences of the human reference genome and available microbial reference sequences into one big reference in one fasta file. The --filter option indicated the position of the first non-human sequence in this big reference.

3. The -mp flag is for including additional mate-pair data, not for standard paired-end data. If you do not have a separate bam file with mate pair data for the same sample, please omit the flag.

4. The genotyping model RANDOM is good for basic analyses. We added the DUP model to correctly genotype insertions that are flanked by duplications as described in the supplement of [Kehr et al. 2017](https://www.nature.com/articles/ng.3801). Note that all insertions that are not flanked by duplications will be incorrectly genotyped using this model. We genotyped the Icelandic data with both models and selected one model for each variant only after imputation.

5. Positions where all samples carry a homozygous genotype are not necessarily false positives, especially not if a clear breakpoint is visible in the bam file. They may, for example, be deletions in the haplotype that was used to make the reference genome. I do not see any relation to the HLA region. I would consider positions where PopIns outputs a heterozygous genotype for all samples suspicious (possible false positives). We filtered the Icelandic call set using measures such as the Hardy-Weinberg equilibrium to remove such calls.

Let me know if I missed to reply to any other question.

Birte

**Comment by Jordi-V** [comment]
Hi Brite,

thanks again for your help and answers. Just only few things.  
1- I looking for where I can download the file "metaRef.fa" but I dont find it... in addition, this fa file is generated only for ref genome hg38?? because I need it for hs37d5... 
2- For the filter flag, I 've to write the same as in the paper?? 471?? or I 've to know something...
2- The adapters flag If I sequence all my samples with HiSeq4000 the flag is -a HiSeq??
3- why is important use the --skippedfile?? the file generated is not using in the above steps in the paper... 
4- Even more, If I want to find inserted viruses with popins, I should use the --filter flag???
5- Only to know... because I ran all my samples with the steps below exposed... You advice me ran again all the steps without -mp, If I ran with the -mp option, everything is wrong?


Thanks again for your help, is really useful

Jordi
